content/go/concepts/working-with-json/working-with-json.md

```diff
@@ -0,0 +1,228 @@
+          {"name": "Caesar Salad", "ingredients": ["Romaine Lettuce", "Croutons", "Parmesan Cheese", "Caesar Dressing"]}
+        ]
+    }`
+
+    // We need a variable to store the data decoded from the unmarshaling process.
+    meals := Meals{}
+
+    // To unmarshal the data, we need to provide:
+    // 1. Data in the format of a slice of bits.
+    // 2. A pointer to the location where the data will be stored.
+    if err := json.Unmarshal([]byte(menu), &meals); err != nil {
+        // Handle the errors in case they occur.
+        fmt.Println(err)
+    }
+
+    // In case we need to add a new object to the
+    meal := Meal{
+        Name:        "Pupusas",
+        Ingredients: []string{"Cheese", "Refried beans", "Tomato Sauce", "Rice flour", "Pickled cabbage"},
+    }
+
+    // Add new meal for the menu
+    meals.Meals = append(meals.Meals, meal)
+
+    // Add new meal in line for the menu
+    meals.Meals = append(meals.Meals, Meal{Name: "Sushi", Ingredients: []string{"Rice", "Fish", "Seaweed", "Wasabi", "Soy Sauce", "Ginger"}})
+
+    // List the menu of meals
+    for _, meal := range meals.Meals {
+        fmt.Printf("---\nMeal: %s\ningredients: %s \n", meal.Name, strings.Join(meal.Ingredients, ", "))
+    }
+
+    // Serialize the object to a slice of bytes
+    data, err := json.Marshal(meals)
+    if err != nil {
+        fmt.Println(err)
+    }
+    fmt.Println("-------------")
+    fmt.Println(string(data))
+
+    // Crate an empty file
+    file, err := os.Create("meals.json")
+    if err != nil {
+        fmt.Println("Error opening file:", err)
+        return
+    }
+
+    defer file.Close()
+
+    // Write data into the file previously created
+    if _, err = file.Write(data); err != nil {
+        fmt.Println(err)
+        return
+    }
+
+}
+```
+
+## NewEncoder and NewDecoder
+
+In Go `json.NewDecoder` is used to format data received from a REST API. Other function that is commonly used is `json.NewEncoder` to exchange data between backend side and external clients that require a JSON response.
+
+In the following example, we will demonstrate the usage of both functions. We will make a call to an external REST API and enhance the result by adding products made from the fruit specified in the query parameter. For example: `http://localhost:4444/fruit?name=Strawberry`
+
+```go
+package main
+
+import (
+    "encoding/json"
+    "errors"
+    "fmt"
+    "io"
+    "log"
+    "net/http"
+    "strings"
+)
+
+type ProductsByFruit map[string][]string
+
+type Fruit struct {
+    Name     string   `json:"name"`
+    ID       int      `json:"id"`
+    Family   string   `json:"family"`
+    Order    string   `json:"order"`
+    Genus    string   `json:"genus"`
+    Products []string `json:"products"`
+}
+
+func main() {
+    mux := http.NewServeMux()
+    mux.HandleFunc("POST /fruit", func(w http.ResponseWriter, r *http.Request) {
```

Comment:
The `mux.HandleFunc` method should not include the HTTP method (`POST`) in the path. The path should just specify the URL pattern, and the HTTP method should be handled separately using `http.MethodPost`.

Fix:

```go
mux.HandleFunc("/fruit", func(w http.ResponseWriter, r *http.Request) {
    if r.Method != http.MethodPost {
        http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
        return
    }
    // Continue handling the request...
})
```